modules/packages/LinearAlgebra.chpl

```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
+
+  for i in 1..n { 
+
+    var max = A[i,i], swaprow = i;
+    for row in (i+1)..n {
+      if (abs(A[row,i]) > abs(max)) {
+        max = A[row,i];
+        swaprow = row;
+      }
+    }
+    if (swaprow != i) {
+      A[i,..] <=> A[swaprow,..];
+      L[i,..] <=> L[swaprow,..];
+      ipiv[i] <=> ipiv[swaprow];
+      numSwap+=1;
+    }
+
+    forall k in i..n {
+      var sum = + reduce (L[i,..] * U[..,k]);
+      U[i,k] = A[i,k] - sum;
+    }
+
+    L[i,i] = 1;
+
+    forall k in (i+1)..n {
+      var sum = + reduce (L[k,..] * U[..,i]);
+      L[k,i] = (A[k,i] - sum) / U[i,i];
+    }
+  } 
+
+  LU = L + U;
+  forall i in 1..n {
+    LU(i,i) = U(i,i);
+  }
+
+  return (LU,ipiv,numSwap);
+}
+
+/*
+  Compute an LU factorization of square matrix `A` 
+  using partial pivoting, such that `A = P * L * U` where P
+  is a permutation matrix. 
+  
+  `L` and `U` are stored in the same matrix `LU` where 
+  the unit diagonal elements of L are not stored.
+  
+  `ipiv` contains the pivot indices such that row i of `A` 
+  was interchanged with row `ipiv(i)`.
+  
+*/
+
+proc lu (A: [?Adom] ?eltType) {
+  if Adom.rank != 2 then
+    halt("Wrong rank for LU factorization");
+
+  if Adom.shape(1) != Adom.shape(2) then
+    halt("LU factorization only supports square matrices");
+
+  var (LU, ipiv, numSwap) = _lu(A);
+  return (LU,ipiv);
+}
+
+proc permute (ipiv, A: [?Adom] ?eltType, transpose=false) {
+  const n = Adom.shape(1);
+  
+  var B: [Adom] eltType;
+  
+  if Adom.rank == 1 {
+    if transpose {
+      forall i in 1..n {
+        B[i] = A[ipiv(i)];
+      }
+    }
+    else {
+      forall i in 1..n {
+        B[ipiv(i)] = A[i];
+      }
+    }
+  }
+  else if Adom.rank == 2 {
+    if transpose {
+      forall i in 1..n {
+        B[i,..] = A[ipiv(i),..];
+      }
+    }
+    else {
+      forall i in 1..n {
+        B[ipiv(i),..] = A[i,..];
+      }
+    }
+  }
+  return B;
+}
+
+/* Return the determinant of a square matrix.
+
+    .. note::
+
+      This procedure performs LU factorization to compute the 
+      determinant. In certain cases, e.g. having a lower/upper
+      triangular matrix, it is more desirable to compute the 
+      determinant manually.
+*/
+
+proc det (A: [?Adom] ?eltType) {
+  if Adom.rank != 2 then
+    halt("Wrong rank for computing determinant");
+
+  if Adom.shape(1) != Adom.shape(2) then
+    halt("Determinant can only be computed from square matrices");
+
+  var (LU,ipiv,numSwap) = _lu(A);
+  const pdet = if numSwap % 2 == 0 then 1 else -1;
+
+  // L[i,i] always = 1, so we only need to take the 
+  // diagonal product of U
+
+  return (* reduce [i in Adom.dim(1)] LU[i,i]) * pdet;
+}
+
+/* Return the solution ``x`` to the linear system `` L * x = b `` 
+    where ``L`` is a lower triangular matrix. `unit_diag` will set the diagonal
+    elements to 1 within this procedure.
+*/
+proc solve_tril (const ref L: [?Ldom] ?eltType, const ref b: [?bdom] eltType, 
+                  unit_diag = true) {
```

Comment:
I don't quite understand the purpose of `unit_diag`. The explanation in the header comment, its use in the implementation, lack of it in `solve_triu` and the similar name but different behavior to scipy.linalg.solve_triangular's `unit_diagonal`. Is it supposed to be an optimization flag for cases where L is unit lower triangular?

Reply:
It is intended to function like scipy.linalg.solve_triangular's unit_diagonal, I will change it accordingly.